lib/fluent/plugin/output.rb

```diff
@@ -800,7 +801,10 @@ def emit_sync(tag, es)
         @counter_mutex.synchronize{ @emit_count += 1 }
         begin
           process(tag, es)
-          @counter_mutex.synchronize{ @emit_records += es.size }
+          @counter_mutex.synchronize do
+            @emit_records += es.size
+            @emit_size += es.to_msgpack_stream.bytesize
```

Comment:
I'm being afraid performance regression of output plugins due to this.
I think it 's too high cost just only for getting stream size even though users might not need this information in most cases.
This patch doesn't provide a way to disable it for output plugins (it can be disabled for input plugins though).

I've measured the performance regression with the following patch and benchmark code:

```diff
diff --git a/lib/fluent/plugin/output.rb b/lib/fluent/plugin/output.rb
index 7d0e3500..e3bd6026 100644
--- a/lib/fluent/plugin/output.rb
+++ b/lib/fluent/plugin/output.rb
@@ -171,9 +171,11 @@ module Fluent
       attr_accessor :output_enqueue_thread_waiting, :dequeued_chunks, :dequeued_chunks_mutex
       # output_enqueue_thread_waiting: for test of output.rb itself
       attr_accessor :retry_for_error_chunk # if true, error flush will be retried even if under_plugin_development is true
+      attr_accessor :use_size_metric
 
       def initialize
         super
+        @use_size_metric = true
         @counter_mutex = Mutex.new
         @buffering = false
         @delayed_commit = false
@@ -803,7 +805,7 @@ module Fluent
           process(tag, es)
           @counter_mutex.synchronize do
             @emit_records += es.size
-            @emit_size += es.to_msgpack_stream.bytesize
+            @emit_size += es.to_msgpack_stream.bytesize if @use_size_metric
           end
         rescue
           @counter_mutex.synchronize{ @num_errors += 1 }
@@ -972,7 +974,7 @@ module Fluent
         end
         @counter_mutex.synchronize do
           @emit_records += records
-          @emit_size += es.to_msgpack_stream.bytesize
+          @emit_size += es.to_msgpack_stream.bytesize if @use_size_metric
         end
         true
       end
@@ -992,7 +994,7 @@ module Fluent
         end
         @counter_mutex.synchronize do
           @emit_records += records
-          @emit_size += es.to_msgpack_stream.bytesize
+          @emit_size += es.to_msgpack_stream.bytesize if @use_size_metric
         end
         true
       end
@@ -1020,7 +1022,7 @@ module Fluent
         end
         @counter_mutex.synchronize do
           @emit_records += records
-          @emit_size += es.to_msgpack_stream.bytesize
+          @emit_size += es.to_msgpack_stream.bytesize if @use_size_metric
         end
         true
       end
```

```ruby
#!/usr/bin/env ruby

require 'benchmark/ips'

require_relative 'test/helper'
require 'fluent/plugin/output'
require 'fluent/event'

class DummyOutput < Fluent::Plugin::Output
  def process(tag, es)
    nil
  end
end

def create_output
  o = DummyOutput.new
  o.configure(config_element())
  o
end

o1 = create_output
o1.use_size_metric = false

o2 = create_output

t = event_time()
es = Fluent::ArrayEventStream.new([ [t, {"key" => "value1"}], [t, {"key" => "value2"}] ])

Benchmark.ips do |x|
  x.report("before") do
    o1.emit_events('tag', es)
  end

  x.report("after") do
    o2.emit_events('tag', es)
  end
end
```

Result:
```
Warming up --------------------------------------
              before   215.850k i/100ms
               after    18.082k i/100ms
Calculating -------------------------------------
              before      2.124M (± 0.5%) i/s -     10.792M in   5.082330s
               after    181.924k (± 0.7%) i/s -    922.182k in   5.069309s
```

It's about 10 times slower than before.
Of course, I think it's too malicious comparison because actual performance is determined by other various aspects and probably it's not bottleneck. Probably it's not so behind in actual cases.

But the performance regression actually exists.
Do you have any concerns about it?

Reply:
> This patch doesn't provide a way to disable it for output plugins (it can be disabled for input plugins though).

Oh, it's not good for us. I think that providing disabling to measure way on output plugin or system config is reasonable for our use case.

I'll implement disable event size measurement system config parameter.

Reply:
BTW, I don't have accurate benchmark yet, but this regression is more sensitive for high frequency loaded environment such as processing over 5000 ~ 6000 events per second case.

Reply:
Fixed in https://github.com/fluent/fluentd/pull/3411/commits/9b13ab283f9e1f3a7b457a0eaa613534f72ec8a0, https://github.com/fluent/fluentd/pull/3411/commits/e400ee6eb460d8f6153f6a207d446ba4c92d0a8c, and https://github.com/fluent/fluentd/pull/3411/commits/c73a83dd8555cbe61ea2604164cf0d3375338665.

I'd confirmed that record size metrics does not update when specifying `enable_size_metrics` as `false` with the following configuration:

```aconf
<system>
  enable_input_metrics true
  enable_size_metrics true
</system>
<source>
  @type monitor_agent
  bind 0.0.0.0
  port 24220
  tag monitor.agents
  emit_interval 5
</source>
<source>
  @type sample
  tag sample.raw
  rate 10
</source>

<match /(monitor|sample).*/>
  @type null
</match>
```

Reply:
Increasing what should be needed to specify enabling size of each record and number of records is no problem for our use case.